python/ray/experimental/channel/torch_tensor_nccl_channel.py

```diff
@@ -138,63 +178,130 @@ def write(self, value: Any):
-            # the task is allowed to return at most one tensor, and its shape
-            # and dtype must match the declared type. Wrap the tensor in a
-            # list since the following calls expect a list.
-            tensors = [tensors]
-
-        self.serialization_ctx.reset_tensors(tensors)
-        data = self._cpu_data_channel.begin_read()
-        self.serialization_ctx.reset_tensors([])
+        """
+        Read a value that may contain torch.Tensors sent via external
+        transport.
+
+        This method:
+        1) Receives torch.Tensors via the tensor data channel (e.g., NCCL).
+        2) Reads the serialized non-tensor data.
+        3) Deserializes the non-tensor data. During deserialization, replaces
+        all found placeholders with the received torch.Tensors.
+
+        If static_non_tensor_data=True was specified, then we only perform step
+        (2) on the first `read` call. Subsequent messages will reuse the
+        first received non-tensor data.
+        """
+        # First, read the tensor data.
+        tensors = self._tensor_data_channel.begin_read()
+        self.serialization_ctx.reset_out_of_band_tensors(tensors)
+
+        # Next, get the serialized non-tensor data.
+        serialized_non_tensor_data = self._serialized_non_tensor_data
+        if serialized_non_tensor_data is None:
+            serialized_non_tensor_data = self._non_tensor_data_channel.begin_read(
+                deserialize=False
+            )
+            if self._static_non_tensor_data:
+                # Cache the serialized non-tensor data for future messages.
+                self._serialized_non_tensor_data = serialized_non_tensor_data
+
+        # Finally, deserialize the non-tensor data. The registered custom
+        # deserializer will replace the found tensor placeholders with
+        # `tensors`. We get index 0 because deserialize_objects expects a list
+        # of objects to deserialize, but we only have one.
+        data = self._worker.deserialize_objects(*serialized_non_tensor_data)[0]
+
+        # Check that all placeholders had a corresponding tensor.
+        (
+            _,
+            deserialized_tensor_placeholders,
+        ) = self.serialization_ctx.reset_out_of_band_tensors([])
+        assert deserialized_tensor_placeholders == set(range(len(tensors)))
 
         return data
 
     def end_read(self) -> None:
-        self._gpu_data_channel.end_read()
-        if self._cpu_data_channel:
-            self._cpu_data_channel.end_read()
+        self._tensor_data_channel.end_read()
+
+        if not self._static_non_tensor_data:
+            # If the non-tensor data is static, then we reuse the serialized
+            # data in the shared memory buffer for all future channel messages.
+            # Therefore, we should only release the buffer if the non-tensor
+            # data is dynamic.
+            self._non_tensor_data_channel.end_read()
```

Comment:
This looks incompatible and will introduce complication for `CompiledDAGRef` API?

Reply:
Should be okay, I'll update the code to just hold the reference instead of calling end_read.